### ISSUE TITLE
MacOS CMake issues

### ISSUE THREAD

**Issue by jeffhammond** [illuhad/hipSYCL#228]
Prompted by https://github.com/illuhad/hipSYCL/issues/222#issuecomment-601104011, I tried to build on Mac but it seems to ignore my explicit specification of Clang location and instead finds `/usr/bin/clang++`.

It is possible that the issue here is that the Homebrew installs of LLVM don't have the required components, but CMake is not doing a proper job detecting that.

```sh
jrhammon-mac02:build jrhammon$ cmake .. -DCMAKE_INSTALL_PREFIX=/opt/hipsycl -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang -DDISABLE_LLVM_VERSION_CHECK=ON
-- The C compiler identification is Clang 9.0.1
-- The CXX compiler identification is Clang 9.0.1
-- Check for working C compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang
-- Check for working C compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang++
-- Check for working CXX compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:17 (set):
  implicitly converting 'INTEGER' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Could NOT find LLVM (missing: LLVM_DIR)
-- Building hipSYCL against LLVM configured from LLVM_DIR-NOTFOUND
-- Selecting clang: /usr/bin/clang++
CMake Error at CMakeLists.txt:77 (message):
  clang include path CLANG_INCLUDE_PATH-NOTFOUND does not exist.  Please
  provide clang's internal include path manually.


-- Using clang include directory: CLANG_INCLUDE_PATH-NOTFOUND
-- Configuring incomplete, errors occurred!
See also "/Users/jrhammon/Work/SYCL/hipSYCL/build/CMakeFiles/CMakeOutput.log".
```

**Comment by illuhad** [comment]
You may need to set `-DLLVM_DIR=/usr/local/Cellar/llvm/9.0.1/lib/cmake` (please verify that this is the directory actually containing the LLVM cmake files - they should be *somewhere* in a directory similar to this one)

**Comment by jeffhammond** [comment]
I can build by removing broken CMake checks and switching a linker flag that isn't supported on MacOS.
```
jrhammon-mac02:build jrhammon$ git clean -dfx ; cmake .. -DCMAKE_INSTALL_PREFIX=/opt/hipsycl -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang -DCLANG_INCLUDE_PATH=/usr/local/Cellar/llvm/9.0.1/include -DCLANG_EXECUTABLE_PATH=/usr/local/Cellar/llvm/9.0.1/bin -DLLVM_BIN_DIR=/usr/local/Cellar/llvm/9.0.1/bin -DCMAKE_CXX_FLAGS="-fopenmp -L/usr/local/Cellar/llvm/9.0.1/lib -I/usr/local/Cellar/llvm/9.0.1/include" && make
jrhammon-mac02:build jrhammon$ git diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 569516d..15038e5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -20,31 +20,6 @@ FORCE)
   endif()
 endif()
 
-if(DISABLE_LLVM_VERSION_CHECK)
-  # just use whatever we are given - some llvm distributions do
-  # not correctly advertise their versions, and if users explicitly point
-  # us to them, let's assume they know what they are doing.
-  find_package(LLVM CONFIG)
-else()
-  # By default, try to find the newest supported LLVM version
-  find_package(LLVM 10 CONFIG QUIET)
-  if(NOT LLVM_FOUND)
-    message(STATUS "No suitable LLVM 10 installation found.")
-    find_package(LLVM 9 CONFIG QUIET)
-    if(NOT LLVM_FOUND)
-      message(STATUS "No suitable LLVM 9 installation found.")
-      find_package(LLVM 8 CONFIG QUIET)
-    endif()
-  endif()
-  if(NOT LLVM_FOUND)
-    message(SEND_ERROR "Could not find an LLVM installation of version 8 or newer. 
-      If you are sure to have passed a suitable LLVM path with -DLLVM_DIR, try again with 
-      -DDISABLE_LLVM_VERSION_CHECK=ON.")
-  endif()
-endif()
-message(STATUS "Building hipSYCL against LLVM configured from ${LLVM_DIR}")
-#find_package(Clang REQUIRED)
-
 # Check for CUDA/ROCm and clang
 find_package(CUDA QUIET)
 # We currently search for hipcc to check for ROCm installation
diff --git a/src/compiler/CMakeLists.txt b/src/compiler/CMakeLists.txt
index 2b4fbfd..e256b2b 100644
--- a/src/compiler/CMakeLists.txt
+++ b/src/compiler/CMakeLists.txt
@@ -31,7 +31,7 @@ endif()
 
 target_link_libraries(hipSYCL_clang
   ${LLVM_LIBS}
-  ${Boost_LIBRARIES} -Wl,-znodelete)
+  ${Boost_LIBRARIES} -Wl,-undefined -Wl,dynamic_lookup) # -Wl,-znodelete)
 
 install(TARGETS hipSYCL_clang
         LIBRARY DESTINATION lib
```

**Comment by jeffhammond** [comment]
Attempting with `LLVM_DIR` active...

```
jrhammon-mac02:build jrhammon$ git clean -dfx ; cmake .. -DCMAKE_INSTALL_PREFIX=/opt/hipsycl -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang -DCLANG_INCLUDE_PATH=/usr/local/Cellar/llvm/9.0.1/include -DCLANG_EXECUTABLE_PATH=/usr/local/Cellar/llvm/9.0.1/bin -DLLVM_BIN_DIR=/usr/local/Cellar/llvm/9.0.1/bin -DCMAKE_CXX_FLAGS="-fopenmp -L/usr/local/Cellar/llvm/9.0.1/lib -I/usr/local/Cellar/llvm/9.0.1/include" -DLLVM_DIR=/usr/local/Cellar/llvm/9.0.1/lib/cmake && make
warning: failed to remove ./: Invalid argument
Removing ./CMakeFiles
Removing ./Makefile
Removing ./cmake_install.cmake
Removing ./hipsycl-config.cmake
Removing ./CMakeCache.txt
Removing ./syclcc.json
Removing ./src
-- The C compiler identification is Clang 9.0.1
-- The CXX compiler identification is Clang 9.0.1
-- Check for working C compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang
-- Check for working C compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang++
-- Check for working CXX compiler: /usr/local/Cellar/llvm/9.0.1/bin/clang++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) at CMakeLists.txt:17 (set):
  implicitly converting 'INTEGER' to 'STRING' type.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- No suitable LLVM 10 installation found.
-- No suitable LLVM 9 installation found.
CMake Error at CMakeLists.txt:40 (message):
  Could not find an LLVM installation of version 8 or newer.

        If you are sure to have passed a suitable LLVM path with -DLLVM_DIR, try again with 
        -DDISABLE_LLVM_VERSION_CHECK=ON.


-- Building hipSYCL against LLVM configured from LLVM_DIR-NOTFOUND
-- Selecting clang: /usr/local/Cellar/llvm/9.0.1/bin
-- Using clang include directory: /usr/local/Cellar/llvm/9.0.1/include
-- Configuring incomplete, errors occurred!
See also "/Users/jrhammon/Work/SYCL/hipSYCL/build/CMakeFiles/CMakeOutput.log".
jrhammon-mac02:build jrhammon$ ll /usr/local/Cellar/llvm/9.0.1/lib/cmake/
total 0
drwxr-xr-x   5 jrhammon  staff   160B Dec 11 11:15 clang
drwxr-xr-x  31 jrhammon  staff   992B Feb 11 07:46 llvm
drwxr-xr-x   4 jrhammon  staff   128B Feb 11 07:46 polly
```

**Comment by illuhad** [comment]
That looks better, additionally try `-DDISABLE_LLVM_VERSION_CHECK=ON` as it suggests. This will disable LLVM version checking. Sometimes cmake fails to detect the version correctly.

**Comment by jeffhammond** [comment]
That option does nothing.  I don't know why, but it was easier to just delete that part of the CMake file.

**Comment by jeffhammond** [comment]
The build I got (details at the bottom) is not able to find `math.h` stuff.  I think it's a simple fix to find some LLVM headers, but I don't know which ones.
```
jrhammon-mac02:Cxx11 jrhammon$ make sycl
/opt/hipSYCL/bin/syclcc-clang -DPRKVERSION="2.16" -std=c++17 -O3 -DHIPSYCL --hipsycl-platform=cpu -Wl,-rpath=/opt/hipSYCL/llvm/lib -I/usr/local/Cellar/llvm/9.0.1/include -I/usr/local/Cellar/boost/1.72.0/include -DBOOST_COMPUTE_USE_CPP11 -DUSE_2D_INDEXING=0 nstream-sycl.cc -o nstream-sycl
In file included from nstream-sycl.cc:65:
In file included from ./prk_sycl.h:7:
In file included from /opt/hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:55:
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:144:1: error: no member named 'exp10f' in the global namespace; did you mean 'exp10'?
HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION(exp10)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:92:48: note: expanded from macro 'HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION'
  HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(func, :: HIPSYCL_PP_CONCATENATE(func,f), ::func) \
                                               ^~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:44:82: note: expanded from macro 'HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD'
  HIPSYCL_KERNEL_TARGET inline float name(float x) { return HIPSYCL_STD_FUNCTION(float_func)(x); } \
                                                                                 ^~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^~~~~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:144:38: note: 'exp10' declared here
HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION(exp10)
                                     ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:144:1: error: no member named 'exp10' in the global namespace; did you mean simply 'exp10'?
HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION(exp10)
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:92:83: note: expanded from macro 'HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION'
  HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(func, :: HIPSYCL_PP_CONCATENATE(func,f), ::func) \
                                                                                  ^~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:45:84: note: expanded from macro 'HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD'
  HIPSYCL_KERNEL_TARGET inline double name(double x) { return HIPSYCL_STD_FUNCTION(double_func)(x); }
                                                                                   ^~~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^~~~~~~~~~~~~
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:144:38: note: 'exp10' declared here
HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION(exp10)
                                     ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:284:63: error: use of undeclared identifier '__cosf'
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(cos, sycl::cos, __cosf);
                                                              ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:285:63: error: use of undeclared identifier '__expf'
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(exp, sycl::exp, __expf);
                                                              ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:287:67: error: use of undeclared identifier '__log10f'; did you mean 'log10f'?
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log10, sycl::log10, __log10f);
                                                                  ^~~~~~~~
                                                                  log10f
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:253:33: note: expanded from macro 'HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION'
  { return HIPSYCL_STD_FUNCTION(fast_sp_func)(x); }
                                ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/math.h:376:14: note: 'log10f' declared here
extern float log10f(float);
             ^
In file included from nstream-sycl.cc:65:
In file included from ./prk_sycl.h:7:
In file included from /opt/hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:55:
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:288:65: error: use of undeclared identifier '__log2f'; did you mean 'log2f'?
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log2, sycl::log2, __log2f);
                                                                ^~~~~~~
                                                                log2f
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:253:33: note: expanded from macro 'HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION'
  { return HIPSYCL_STD_FUNCTION(fast_sp_func)(x); }
                                ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/math.h:380:14: note: 'log2f' declared here
extern float log2f(float);
             ^
In file included from nstream-sycl.cc:65:
In file included from ./prk_sycl.h:7:
In file included from /opt/hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:55:
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:289:63: error: use of undeclared identifier '__logf'; did you mean 'logf'?
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log, sycl::log, __logf);
                                                              ^~~~~~
                                                              logf
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:253:33: note: expanded from macro 'HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION'
  { return HIPSYCL_STD_FUNCTION(fast_sp_func)(x); }
                                ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/math.h:372:14: note: 'logf' declared here
extern float logf(float);
             ^
In file included from nstream-sycl.cc:65:
In file included from ./prk_sycl.h:7:
In file included from /opt/hipSYCL/bin/../include/CL/sycl.hpp:31:
In file included from /opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/sycl.hpp:55:
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:290:63: error: use of undeclared identifier '__sinf'
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(sin, sycl::sin, __sinf);
                                                              ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:291:63: error: use of undeclared identifier '__tanf'
HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(tan, sycl::tan, __tanf);
                                                              ^
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/math.hpp:303:66: error: use of undeclared identifier '__powf'; did you mean 'powf'?
inline float pow(float x, float y) { return HIPSYCL_STD_FUNCTION(__powf)(x,y); }
                                            ~~~~~~~~~~~~~~~~~~~~~^~~~~~~
                                            powf
/opt/hipSYCL/bin/../include/CL/../hipSYCL/sycl/builtin.hpp:60:45: note: expanded from macro 'HIPSYCL_STD_FUNCTION'
#define HIPSYCL_STD_FUNCTION(function_name) function_name
                                            ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include/math.h:428:14: note: 'powf' declared here
extern float powf(float, float);
             ^
10 errors generated.
make: *** [nstream-sycl] Error 1
```

```
jrhammon-mac02:build jrhammon$ git diff 
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 569516d..06762dc 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -25,22 +25,6 @@ if(DISABLE_LLVM_VERSION_CHECK)
   # not correctly advertise their versions, and if users explicitly point
   # us to them, let's assume they know what they are doing.
   find_package(LLVM CONFIG)
-else()
-  # By default, try to find the newest supported LLVM version
-  find_package(LLVM 10 CONFIG QUIET)
-  if(NOT LLVM_FOUND)
-    message(STATUS "No suitable LLVM 10 installation found.")
-    find_package(LLVM 9 CONFIG QUIET)
-    if(NOT LLVM_FOUND)
-      message(STATUS "No suitable LLVM 9 installation found.")
-      find_package(LLVM 8 CONFIG QUIET)
-    endif()
-  endif()
-  if(NOT LLVM_FOUND)
-    message(SEND_ERROR "Could not find an LLVM installation of version 8 or newer. 
-      If you are sure to have passed a suitable LLVM path with -DLLVM_DIR, try again with 
-      -DDISABLE_LLVM_VERSION_CHECK=ON.")
-  endif()
 endif()
 message(STATUS "Building hipSYCL against LLVM configured from ${LLVM_DIR}")
 #find_package(Clang REQUIRED)
diff --git a/src/compiler/CMakeLists.txt b/src/compiler/CMakeLists.txt
index 2b4fbfd..b18b2e0 100644
--- a/src/compiler/CMakeLists.txt
+++ b/src/compiler/CMakeLists.txt
@@ -29,9 +29,17 @@ if(NOT ${LLVM_ENABLE_RTTI})
   target_compile_options(hipSYCL_clang PRIVATE -fno-rtti)
 endif()
 
+if(APPLE)
+  set (CMAKE_MODULE_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -Wl,-flat_namespace -Wl,-undefined,suppress")
+  #set (CMAKE_MODULE_LINK_FLAGS "${CMAKE_CXX_LINK_FLAGS} -Wl,-undefined -Wl,dynamic_lookup")
+else()
+  set (CMAKE_MODULE_LINKER_FLAGS "${CMAKE_CXX_LINK_FLAGS} -Wl,-znodelete")
+endif(APPLE)
+
 target_link_libraries(hipSYCL_clang
   ${LLVM_LIBS}
-  ${Boost_LIBRARIES} -Wl,-znodelete)
+  ${Boost_LIBRARIES}
+  ${CMAKE_MODULE_LINK_FLAGS})
 
 install(TARGETS hipSYCL_clang
         LIBRARY DESTINATION lib
jrhammon-mac02:build jrhammon$ git clean -dfx ; cmake .. -DCMAKE_INSTALL_PREFIX=/opt/hipsycl -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang++ -DCMAKE_C_COMPILER=/usr/local/Cellar/llvm/9.0.1/bin/clang -DCLANG_INCLUDE_PATH=/usr/local/Cellar/llvm/9.0.1/include -DCLANG_EXECUTABLE_PATH=/usr/local/Cellar/llvm/9.0.1/bin -DLLVM_BIN_DIR=/usr/local/Cellar/llvm/9.0.1/bin -DCMAKE_CXX_FLAGS="-fopenmp -Wl,-L/usr/local/Cellar/llvm/9.0.1/lib -Wno-unused-command-line-argument -I/usr/local/Cellar/llvm/9.0.1/include" && make
```

**Comment by mrzv** [comment]
With your magical `CMAKE_MODULE_LINKER_FLAGS` changes, everything compiles! We'll see if it actually runs.

**Comment by mrzv** [comment]
And I also get problems with math, when I try to use `syclcc-clang`.

**Comment by mrzv** [comment]
Here's the diff for `include/hipSYCL/sycl/math.h` that makes a test program compile:
```
90a91,93
> //#define HIPSYCL_DEFINE_GENFLOAT_STD_FUNCTION(func) \
> //  HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(func, :: HIPSYCL_PP_CONCATENATE(func,f), ::func) \
> //  HIPSYCL_DEFINE_FLOATN_MATH_FUNCTION(func, func)
92c95
<   HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(func, :: HIPSYCL_PP_CONCATENATE(func,f), ::func) \
---
>   HIPSYCL_DEFINE_FLOATING_POINT_OVERLOAD(func, HIPSYCL_PP_CONCATENATE(func,), func) \
284,285c287,288
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(cos, sycl::cos, __cosf);
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(exp, sycl::exp, __expf);
---
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(cos, sycl::cos, cosf);
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(exp, sycl::exp, expf);
287,291c290,294
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log10, sycl::log10, __log10f);
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log2, sycl::log2, __log2f);
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log, sycl::log, __logf);
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(sin, sycl::sin, __sinf);
< HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(tan, sycl::tan, __tanf);
---
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log10, sycl::log10, log10f);
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log2, sycl::log2, log2f);
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(log, sycl::log, logf);
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(sin, sycl::sin, sinf);
> HIPSYCL_DEFINE_FAST_SINGLE_PRECISION_FUNCTION(tan, sycl::tan, tanf);
303c306
< inline float pow(float x, float y) { return HIPSYCL_STD_FUNCTION(__powf)(x,y); }
---
> inline float pow(float x, float y) { return HIPSYCL_STD_FUNCTION(powf)(x,y); }
```
But now I'm getting:
```
/opt/hipSYCL/bin/syclcc-clang --hipsycl-platform=cpu -O3 example.cpp -o example-cpu
ld: library not found for -lomp
clang-9: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by jeffhammond** [comment]
`-DCMAKE_CXX_FLAGS="-fopenmp -L/usr/local/Cellar/llvm/9.0.1/lib -I/usr/local/Cellar/llvm/9.0.1/include"` should fix your OpenMP issue, assuming you modify the LLVM path accordingly.

**Comment by mrzv** [comment]
Yes! Passing those flags to `syclcc-clang` worked. It would be nice if it was possible to hard-code these into `syclcc-clang` itself.

Now the actual compiled code gives an error at the very end of its run:
```
libc++abi.dylib: terminating with uncaught exception of type std::__1::system_error: mutex lock failed: Invalid argument
```
But that's probably a subject for a different issue.

**Comment by illuhad** [comment]
@mrzv `syclcc-clang` passes all flags that should be required, including `-fopenmp`. The issues that have come up here are beyond me and the fixes presented do not address the root cause of *something* being very broken in that toolchain...

**Comment by mrzv** [comment]
@illuhad I get your larger point. I just want to say that I think macOS support is a worthwhile goal, even given the state of CUDA and ROCm there.

@jeffhammond Have you tried to get hipSYCL to install with Spack? I see that a package exists, but I haven't been able to compile LLVM with Spack on any platform (not Linux, not macOS).

**Comment by illuhad** [comment]
@mrzv Sure, I understand that it can be useful for development or debugging before deploying on e.g. a Linux-based GPU machine or cluster. In the end, it's an open source project, so if someone is willing to test and maintain it on Mac (which I cannot do because of lack of test machine) I will happily merge PRs that may be required - as long as the required changes are within reason, of course :-)

What is your LLVM spack issue? I did a full hipSYCL+cuda spack build on monday and it worked fine. The system was running CentOS 7.

**Comment by mrzv** [comment]
I'm getting the following error, when compiling LLVM:
```
In file included from /tmp/dmitriy/spack-stage/spack-stage-llvm-9.0.1-43f5ss7gfd3zoan4r6zckg25fmrx75hn/spack-src/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:162:
/tmp/dmitriy/spack-stage/spack-stage-llvm-9.0.1-43f5ss7gfd3zoan4r6zckg25fmrx75hn/spack-src/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h:342:72: error: narrowing conversion of '-1' from 'int#
  342 |     typedef char IMPL_PASTE(assertion_failed_##_, line)[2*(int)(pred)-1]
      |       
```

**Comment by illuhad** [comment]
@mrzv Interesting, that looks similar to an issue that @sbalint98 and I had to work around when compiling the ROCm packages - we had assumed it was ROCm specific...@sbalint98, can you check if this is really the same error?

**Comment by sbalint98** [comment]
@illuhad Yes this looks very similar to the problem we have encountered.
@mrzv In our case the reason was probably that arch linux now uses glibc 2.31 which seems to brake some parts of LLVM 9. 
In our case the first error message was the following:
```
/git/llvm-vanilla/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.cc:11
31:1: error: 'assertion_failed__1131' declared as an array with a negative size
CHECK_SIZE_AND_OFFSET(ipc_perm, mode);
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/git/llvm-vanilla/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h:143
7:3: note: expanded from macro 'CHECK_SIZE_AND_OFFSET'
  COMPILER_CHECK(sizeof(((__sanitizer_##CLASS *) NULL)->MEMBER) == \
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/git/llvm-vanilla/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h:336:30: not
e: expanded from macro 'COMPILER_CHECK'ísísd
#define COMPILER_CHECK(pred) IMPL_COMPILER_ASSERT(pred, __LINE__)
                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/git/llvm-vanilla/compiler-rt/lib/sanitizer_common/sanitizer_internal_defs.h:342:57: not
e: expanded from macro 'IMPL_COMPILER_ASSERT'
    typedef char IMPL_PASTE(assertion_failed_##_, line)[2*(int)(pred)-1]
```
Removing the CHECK_SIZE_AND_OFFSET(ipc_perm, mode) seems to resolve the problem. According to my understanding this what the [patch](https://github.com/llvm/llvm-project/commit/947f9692440836dcb8d88b74b69dd379d85974ce.patch) does as well

After our discussion with @illuhad I have found a [bug report ](https://bugs.gentoo.org/708430) and a [patch](https://github.com/llvm/llvm-project/commit/947f9692440836dcb8d88b74b69dd379d85974ce.patch) regarding this problem.

**Comment by mrzv** [comment]
@sbalint98 Very interesting. I'm also on Arch Linux. I suppose I can add that patch to my local LLVM formula in spack, and see if that solves the problem.

How does Arch fix this in its own LLVM package? Do you see anything obvious in the [PKGBUILD](https://git.archlinux.org/svntogit/packages.git/tree/trunk/PKGBUILD?h=packages/llvm)?

**Comment by jeffhammond** [comment]
I will do regular testing once MacOS is supported in some form. We can also use Travis CI, assuming there’s a way to not build LLVM from source every time.

**Comment by mrzv** [comment]
@sbalint98 @illuhad I did not manage to apply that patch, but simply telling spack to use `llvm@master` worked: `spack install hipsycl@master ^llvm@master`. So that's great.

I still hope macOS works out in some form.

**Comment by sbalint98** [comment]
@mrzv Installing LLVM master might not be the best idea, in my opinion. There are some stability problems (see issue: #223 ) with llvm10 already, and probably master is even worse than that. 

By the way, we have updated our repositories, so now there is a functional stack readily available for arch.

**Comment by mrzv** [comment]
@sbalint98 Sorry, which repositories are you talking about?

**Comment by sbalint98** [comment]
I meant the repositories of hipSYCL see the readme in [install/scripts](https://github.com/illuhad/hipSYCL/tree/master/install/scripts#installing-from-repositories) We currently support archlinux, ubuntu 18.04 and centos 7

**Comment by mrzv** [comment]
Oh, I see. The hipSYCL compilation was working fine for me on Arch, but can you fix the AUR package, while you are at it? It's outdated and broken.

**Comment by illuhad** [comment]
@mrzv Those PKGBUILDs are no longer maintained because they require a lot of additional effort. Our current efforts focus on creating a packaging infrastructure where we can reuse as much of the infrastructure as possible across all targeted distributions. This is not easily possible with distribution specific platforms such as the AUR. Additionally, maintaining dedicated AUR ROCm packages just to target Arch/AUR is out of the question for us in terms of the effort required.
There is some interest from the community however to maintain them - this is the path forward in my opinion.

**Comment by mrzv** [comment]
@illuhad Am I understanding it correctly that you are building Arch packages without PKGBUILDs? That's unusual.

Perhaps, you could orphan the AUR packages? Somebody with enough energy would pick them up. In Arch, the first place anyone would look for a new package is AUR. I didn't know there is a dedicated Arch repo for hipSYCL until this thread, but even knowing it, it's unlikely I'd actually use it. I don't think I'm atypical in this regard.

**Comment by illuhad** [comment]
@mrzv Well, we use `makepkg` and PKGBUILDs, however the pkgbuilds are mostly trivial as they simply take an existing installation (created using the installation scripts in `install/scripts`) and wrap them in a package. The entire pipeline is in the `install` directory if you want to have a look.

I had already posted some more information on [the AUR](https://aur.archlinux.org/packages/hipsycl-rocm-git/#comment-731730).

Note that the `hipsycl-cuda` and `hipsycl-rocm` packages are in development and are already maintained by the community (thanks @acxz !).

>  I didn't know there is a dedicated Arch repo for hipSYCL until this thread, but even knowing it, it's unlikely I'd actually use it. I don't think I'm atypical in this regard.

Well, the repositories are the recommended way to install hipSYCL and described as such in the readme ;)
I am aware that Arch Linux users tend to prefer to build things themselves. For hipSYCL I still believe it makes a lot of sense to provide binary packages since
* it removes the burden from users of having to compile a potentially huge stack by themselves, including a custom LLVM stack, as well as at least two more clang/LLVM stacks when also building ROCm.
* it provides us with a consistent stack that we can verify and test against for all distributions.

In general, we could not carry the maintenance burden if we had to adhere to distribution-specific rules, guidelines and different build system if we relied on distribution specific platforms such as the AUR for all distributions we support. In particular because almost no distribution has ROCm packages already, so we'd also need to maintain them as well.

**Comment by mrzv** [comment]
@illuhad Interesting. Thanks for the explanation.

**Comment by acxz** [comment]
Thx for mentioning me! So right now we have done some good work at https://github.com/rocm-arch/rocm-arch to get the rocm stack working on ArchLinux, feel free to try it out and send any bug reports our way!

Since we got the rocm stack working on arch, I decided to create the `hipsycl-cuda` and `hipsycl-rocm` packages (you can also try `hipsycl-cpu`), these should also be working. Eventually the plan is to send a request over to https://github.com/arch4edu/arch4edu to provide binary packages for `hipsycl` on arch and once we get enough votes on the AUR, I can submit a request to have them in the official ArchLinux repositories.

Also while we are discussing this, @illuhad do you mind giving me co-maintainership on the `-git` versions of the packages? Or if you don't plan to maintain them, you can transfer the ownership to me, by adding me as a co-maintainer and then disowning the package.

**Comment by illuhad** [comment]
@acxz Great to see the progress! :) I have given you co-maintainership. I'm not sure how involved I will be with these packages in the future, so potentially I will disown them and give you full ownership at some point.
Note that there may be some changes to the design in hipSYCL in the future making the distinction between rocm and cuda variants unnecessary - I will keep you informed. In any case, I think it will be a good strategy moving forward to attempt to unify them.
Also, if it doesn't work out with arch4edu, perhaps we can also offer to build hipSYCL and related packages (like ROCm) on our hardware (i.e. Heidelberg university infrastructure). Let me know if you would need this so that I can investigate this option.

**Comment by acxz** [comment]
Sweet, thx so much! Unifying them sounds like a really good idea.

If you would like to host the packages, another option is for you to create a mirror of arch4edu. It also has packages like ROS (Robot Operating System) that others in your university or region might benefit with a closer mirror.
[Here](https://github.com/arch4edu/arch4edu/issues/60) is the issue tracking the inclusion of `rocm` and once that is finished we will request `hipsycl` to be on there as well.

**Comment by ax3l** [comment]
Coming back to @jeffhammond 's original issue: [this tweet](https://twitter.com/illuhad/status/1276559369883312128) got me interested, so I took out my Linux laptop, spun up a macOS image on GitHub action and submitted #271 for macOS support.

The main tweaks:
- make sure to install also `libomp` with `brew` for OpenMP support on macOS.
- the compiler wrapper on macOS needs an extra flag [`-Xclang -fopenmp`](https://gitlab.kitware.com/cmake/cmake/-/blob/v3.17.3/Modules/FindOpenMP.cmake#L99) (or `-Xpreprocessor -fopenmp`) and we have to pass the include directory to `omp.h`
- avoid unknown linker flags [as mentioned above](https://github.com/illuhad/hipSYCL/issues/228#issuecomment-601316674)

**Comment by jeffhammond** [comment]
I’ll test locally ASAP. I’ve been using triSYCL but it’s nice to have options.

**Comment by keryell** [comment]
> I’ll test locally ASAP. I’ve been using triSYCL but it’s nice to have options.

At some point, when everything is streamlined, it would be nice to document how to run triSYCL on a Mac. My manager has a Mac... :-)

**Comment by jeffhammond** [comment]
@keryell See https://github.com/triSYCL/triSYCL/issues/285.
